subprojects/core/src/main/java/org/gradle/initialization/DefaultBuildGateToken.java

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright 2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.initialization;
+
+import com.google.common.collect.Lists;
+import org.gradle.internal.UncheckedException;
+
+import java.util.List;
+import java.util.concurrent.locks.Condition;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+public class DefaultBuildGateToken implements BuildGateToken {
+    private final Lock lock = new ReentrantLock();
+    private final Condition opened = lock.newCondition();
+    private final List<GateKeeper> gatekeepers = Lists.newArrayList();
+    private GateKeeper openedBy;
+
+    @Override
+    public void addGateKeeper(GateKeeper gatekeeper) {
+        lock.lock();
+        try {
+            gatekeepers.add(gatekeeper);
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    @Override
+    public void open(GateKeeper gatekeeper) {
```

Comment:
Yep, I like this a lot better.

If we restrict the number of gate keepers you can have, we'll need to do something about having multiple PlayRun tasks in a single build.